lms/services/assignment.py

```diff
@@ -1,64 +1,183 @@
+            query = query.filter(
+                Assignment.ext_lti_assignment_id == ext_lti_assignment_id
+            )
+        elif resource_link_id and ext_lti_assignment_id:
+            query = query.filter(
+                (
+                    # Regular canvas launch
+                    (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                    # Configuring a file assignment that was stored in the DB
+                    # before all canvas assignments were stored in the DB
+                    | (
+                        (Assignment.resource_link_id == resource_link_id)
+                        & (Assignment.ext_lti_assignment_id.is_(None))
+                    )
+                    #  First launch of a newly configured assignment
+                    | (
+                        (Assignment.resource_link_id.is_(None))
+                        & (Assignment.ext_lti_assignment_id == ext_lti_assignment_id)
+                    )
+                )
+            )
+        else:
+            log.exception(
+                "Can't get an assignment without neither resource_link_id or ext_lti_assignment_id"
+            )
+            return None
+
+        assignments = query.all()
+        if not assignments:
+            return None
+
+        if len(assignments) == 2:
+            assignment = self._merge_canvas_assignments(assignments)
+        else:
+            assignment = assignments[0]
+
+        if resource_link_id and not assignment.resource_link_id:
+            # First lunch of a canvas assignment, fill the resource_link_id now
+            assignment.resource_link_id = resource_link_id
+
+        return assignment
+
     @lru_cache(maxsize=128)
-    def get(self, tool_consumer_instance_guid, resource_link_id):
-        return (
-            self._db.query(Assignment)
-            .filter_by(
-                tool_consumer_instance_guid=tool_consumer_instance_guid,
-                resource_link_id=resource_link_id,
+    def get(
+        self,
+        tool_consumer_instance_guid,
+        resource_link_id=None,
+        ext_lti_assignment_id=None,
+    ):
+        """Get an assignment using using resource_link_id, ext_lti_assignment_id or both."""
+
+        if resource_link_id and not ext_lti_assignment_id:
+            # Non canvas assignments
+            return (
+                self._db.query(Assignment)
+                .filter(
+                    Assignment.tool_consumer_instance_guid
+                    == tool_consumer_instance_guid,
+                    Assignment.resource_link_id == resource_link_id,
+                )
+                .one_or_none()
             )
-            .one_or_none()
+
+        return self._get_canvas(
+            tool_consumer_instance_guid, resource_link_id, ext_lti_assignment_id
         )
```

Comment:
`.get` look less daunting now although all the code is still here (`._get_canvas` and `_merge_canvas_assignments`). I think is a good solution without a more deep refactoring.